src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs

```diff
@@ -951,5 +940,237 @@ public static Task ReplayWarningsAndErrorsAsync(IEnumerable<IAssetsLogMessage> m
+                metadata.Add("VersionRange", versionRange);
+            }
+
+            return new MSBuildItem(item.Identity, metadata);
+        }
+
+        private static bool AddCentralVersionDependencyIfNotExist(PackageSpec spec, NuGetFramework framework, CentralVersionDependency centralVersionDependency)
+        {
+            var frameworkInfo = spec.GetTargetFramework(framework);
+            var frameworkDepIds = frameworkInfo.Dependencies.Select(d => d.Name).ToList();
+            if (!frameworkInfo.CentralVersionDependencies
+                            .Select(d => d.Name)
+                            .Contains(centralVersionDependency.Name, StringComparer.OrdinalIgnoreCase))
+            {
+                // Do not duplicate the item in dependecy and in the central list
+                if (!frameworkDepIds.Contains(centralVersionDependency.Name, StringComparer.OrdinalIgnoreCase))
+                {
+                    frameworkInfo.CentralVersionDependencies.Add(centralVersionDependency);
+                }
+
+                return true;
+            }
+
+            return false;
+        }
+
+        private static void AddCentralVersionDependencies(PackageSpec spec, IEnumerable<IMSBuildItem> items)
+        {
+            var centralVersionsDependencies = CreateCentralVersionDependencies(items, spec.TargetFrameworks);
+            foreach (var framework in centralVersionsDependencies.Keys)
+            {
+                foreach (var centralVersionsDependency in centralVersionsDependencies[framework])
+                {
+                    AddCentralVersionDependencyIfNotExist(spec, framework, centralVersionsDependency);
+                }
+            }
+        }
+
+        internal static IEnumerable<IMSBuildItem> GetPackageReferenceItems(IEnumerable<IMSBuildItem> items, bool cpvmEnabled)
+        {
+            if (cpvmEnabled)
+            {        
+                return MergeCentralPackageVersionsInPackageReferences(packageReferences: GetItemByType(items, "Dependency").ToList(),
+                    centralPackageVersions: GetItemByType(items, "CentralVersionDependency").ToList());
+            }
+
+            return GetItemByType(items, "Dependency");
+        }
+
+        private static void ThrowIfItemsHaveVersion(IEnumerable<IMSBuildItem> items)
+        {
+            var builder = new StringBuilder();
+            foreach(var item in items)
+            {
+                var version = item.GetProperty("Version");
+                var versionRange = item.GetProperty("VersionRange");
+                if (!string.IsNullOrEmpty(version) || !string.IsNullOrEmpty(versionRange))
+                {
+                    builder.Append(item.GetProperty("Id"));
+                }
+            }
+            if(builder.Length > 0)
+            {
+                throw new ArgumentException(string.Format(CultureInfo.CurrentCulture, Strings.Error_CentralPackageVersions_VersionsNotAllowed, builder.ToString()));
```

Comment:
I think that this error needs an NU code, so if customers see it, then can easily find a docs page to help explain the issue.

Reply:
I added the NU code.  Currently the error will be added to the PackageSpec and it is logged in the RestoreCommand. Currently it is not displayed in the UI as it is not added to the assets file. This is a design decision that it will be good to review more carefully. It also may be related with https://github.com/NuGet/Home/issues/7717